scipy/optimize/_hessian_update_strategy.py

```diff
@@ -0,0 +1,338 @@
+"""quasi-Newton update strategies."""
+
+from __future__ import division, print_function, absolute_import
+import numpy as np
+from numpy.linalg import norm
+from scipy.linalg import get_blas_funcs
+
+class HessianUpdateStrategy(object):
+    """Virtual interface for implementing Hessian update strategies.
+
+    Should implement four methods: ``initialize``, ``update``,
+    ``dot`` and ``get_matrix``.
+
+    Notes
+    -----
+    Any instance of a class that implements this interface,
+    can be accepted by the method ``minimize`` and used by
+    the compatible solvers to approximate the Hessian (or
+    inverse Hessian) used by the optimization algorithms.
+    """
+
+    def initialize(self, n, approx_type):
+        """Initialize internal matrix.
+
+        Alocate internal memory for storing and updating
+        the Hessian or its inverse.
+
+        Parameters
+        ----------
+        n : int
+            Problem dimension.
+        approx_type : {'hess', 'inv_hess'}
+            Selects either the Hessian or the inverse Hessian.
+            When set to 'hess' the Hessian will be stored and updated.
+            When set to 'inv_hess' its inverse will be used instead.
+        """
+        raise NotImplementedError("The method ``initialize(n, approx_type)``"
+                                  " is not implemented.")
+
+    def update(self, delta_x, delta_grad):
+        """Update matrix.
+
+        Update Hessian matrix or its inverse (depending on how 'approx_type'
+        is defined) using information about the last evaluated points.
+
+        Parameters
+        ----------
+        delta_x : ndarray
+            The difference between two points the gradient
+            function have been evaluated at: ``delta_x = x2 - x1``.
+        delta_grad : ndarray
+            The difference between the gradients:
+            ``delta_grad = grad(x2) - grad(x1)``.
+        """
+        raise NotImplementedError("The method ``update(delta_x, delta_grad)``"
+                                  " is not implemented.")
+
+    def dot(self, p):
+        """Matrix-vector multiplication.
```

Comment:
"Compute the product of the internal matrix with a given vector."

Reply:
I wrote "Compute internal matrix product with the given vector." I think it a little less verbose, and it seems correct for me.

If you think it is ambiguous or something like that I will rewrote it as you have suggested above.

Reply:
I read somewhere that stacking nouns as adjectives is not always correct and when not sure it's better to avoid it. In this case it is not exactly clear what is "internal matrix product", I would prefer my version honestly.

Reply:
ok, I will adopt what you have written above.